### PR TITLE
Reoriented WellKnownLabels API parameters around generic requirements

### PR DESCRIPTION
**Pull request by ellistarn**
**1. Issue, if available:**

**2. Description of changes:**
Experimental API Change/Proposal. 

### Behavior
- pod.NodeSelector, pod.Affinity.NodeAffinity, provisioner.Labels, and Provisioner.Requirements all communicate using the same well known label semantic and their requirements are intersected.

### Motivation
- Users could not specify a `NotIn` semantic at the provisioner level (e.g. excluding specific instance types)
- Users can use now upstream well known labels at the provisioner level, avoiding multiple concepts for the same thing (e.g. instanceType == node.kubernetes.io/instance-type, it's always the latter now)
- Any additional well known labels no longer require API Changes, a new CRD installation, etc. Just a new controller binary.
- Neutral and vendor specific labels were not modeled in the same way, which wasn't intuitive, since they behave the same.

### Future
- This will require a v1alpha5 bump (in a follow-on PR, before v0.4.1 release)
- It's backwards compatible to implement a sibling `Preferences` field, to configure soft constraints at the provisioner level. I punted on this complexity for this PR. 

### Example Provisioner
```
spec:
  labels: # sugar for requirements w/ hardcoded value
    custom: label
    topology.kubernetes.io/zone: us-west-2a 
  requirements: # flexible In/NotIn syntax for well known labels
  - key: node.k8s.aws/capacity-type
    operator: In
    values: ["spot"]
  - key: node.kubernetes.io/instance-type
    operator: NotIn
    values: ["m5.large", "m5.2xlarge"]
  - key: custom.billing.label
    operator: In
    values: ["myteam"]
  provider:
    cluster:
      endpoint: https://BE5C79AC824BB1EF67DB8CC14D6A263F.gr7.us-west-2.eks.amazonaws.com
      name: etarn
```

**3. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/apis/provisioning/v1alpha4/provisioner.go [hunk 1]
```diff
@@ -100,3 +93,97 @@ type ProvisionerList struct {
 	metav1.ListMeta `json:"metadata,omitempty"`
 	Items           []Provisioner `json:"items"`
 }
+
+// Zones for the constraints
+func (c *Constraints) Zones() []string {
+	return c.Requirements.GetLabelValues(v1.LabelTopologyZone)
+}
+
+/// InstanceTypes for the constraints
```
Comment (JacobGabrielson) [inline pkg/apis/provisioning/v1alpha4/provisioner.go:102]: extra `/`

File: pkg/cloudprovider/registry/register.go [hunk 2]
```diff
@@ -20,7 +20,7 @@ import (
 
 	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha4"
 	"github.com/awslabs/karpenter/pkg/cloudprovider"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (JacobGabrielson) [inline pkg/cloudprovider/registry/register.go:23]: Why is this necessary?

File: pkg/controllers/allocation/scheduling/scheduler.go [hunk 3]
```diff
@@ -101,18 +100,18 @@ func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha4.Provisioner
...
 func (s *Scheduler) solve(ctx context.Context, constraints *v1alpha4.Constraints, pods []*v1.Pod) ([]*Schedule, error) {
-	// Apply runtime constraints
+	// Consolidate requirements in memory before executing scheduling logic.
+	// This is a performance optimization to avoid executing requirement
+	// evaluation logic redundantly for every pod.
 	constraints = constraints.DeepCopy()
-	if err := constraints.Constrain(ctx); err != nil {
-		return nil, fmt.Errorf("applying constraints, %w", err)
-	}
+	constraints.Requirements = constraints.CombineRequirements()
```
Comment (JacobGabrielson) [inline pkg/controllers/allocation/scheduling/scheduler.go:107]: Seems a bit odd for the constraints struct not to handle this itself (as opposed to the Scheduler reaching in and doing it?)

**Review by JacobGabrielson — COMMENTED** [review]
Lgtm overall, just a few minor comments

File: pkg/cloudprovider/registry/register.go (hunk 2 above)
Comment (ellistarn) [reply pkg/cloudprovider/registry/register.go:23]: An epic battle with my linter -- which I lost. Will fix.

File: pkg/controllers/allocation/scheduling/scheduler.go (hunk 3 above)
Comment (ellistarn) [reply pkg/controllers/allocation/scheduling/scheduler.go:107]: In the first example, we would need to pass these combined requirements (along w/ provider, taints, and labels) into the per-pod scheduling logic. By the time we're passing those 4 things, we might as well use the constraints object, since it has those exact fields.

I like the second approach you noted.

File: pkg/cloudprovider/aws/securitygroups.go [hunk 4]
```diff
@@ -47,7 +47,7 @@ func (s *SecurityGroupProvider) Get(ctx context.Context, constraints *v1alpha1.C
 	}
 	// Fail if no security groups found
 	if len(securityGroups) == 0 {
-		return nil, fmt.Errorf("no security groups exist given constraints")
+		return nil, fmt.Errorf("no security groups exist given provider")
```
Comment (bwagner5) [inline pkg/cloudprovider/aws/securitygroups.go:50]: provider -> provisioner?

File: pkg/controllers/allocation/scheduling/scheduler.go [hunk 5]
```diff
@@ -101,18 +100,17 @@ func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha4.Provisioner
...
 func (s *Scheduler) solve(ctx context.Context, constraints *v1alpha4.Constraints, pods []*v1.Pod) ([]*Schedule, error) {
-	// Apply runtime constraints
-	constraints = constraints.DeepCopy()
-	if err := constraints.Constrain(ctx); err != nil {
-		return nil, fmt.Errorf("applying constraints, %w", err)
-	}
+	// Consolidate requirements in memory before executing scheduling logic.
+	// This is a performance optimization to avoid executing requirement
+	// evaluation logic redundantly for every pod.
+	constraints = constraints.Compress()
```
Comment (bwagner5) [inline pkg/controllers/allocation/scheduling/scheduler.go:106]: I like `Consolidate()` more than `Compress()` personally :)

**Review by bwagner5 — DISMISSED** [review]
/lgtm
